flink-python/pyflink/fn_execution/datastream/process/async_function/operation.py

```diff
@@ -0,0 +1,289 @@
+                    self._queue.wait_for_completed_elements()
+            except Exception as e:
+                self._running = False
+                self._exception_handler(e)
+
+    def stop(self):
+        self._running = False
+
+
+class AsyncFunctionRunner(threading.Thread):
+    def __init__(self, exception_handler: Callable[[Exception], None]):
+        super().__init__()
+        self._exception_handler = exception_handler
+        self._loop = None
+
+    def run(self):
+        self._loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self._loop)
+        self._loop.run_forever()
+
+    def stop(self):
+        if self._loop is not None:
+            self._loop.stop()
+            self._loop = None
+
+    async def exception_handler_wrapper(self, async_function, *arg):
+        try:
+            await async_function(*arg)
+        except Exception as e:
+            self._exception_handler(e)
+
+    def run_async(self, async_function, *arg):
+        wrapped_function = self.exception_handler_wrapper(async_function, *arg)
+        asyncio.run_coroutine_threadsafe(wrapped_function, self._loop)
+
+
+class AsyncOperation(Operation):
+    def __init__(self, serialized_fn, operator_state_backend):
+        super(AsyncOperation, self).__init__(serialized_fn, operator_state_backend)
+        (
+            self.class_name,
+            self.open_func,
+            self.close_func,
+            self.async_invoke_func,
+            self.timeout_func,
+            self._timeout,
+            capacity,
+            output_mode
+        ) = extract_async_function(
+            user_defined_function_proto=serialized_fn,
+            runtime_context=StreamingRuntimeContext.of(
+                serialized_fn.runtime_context, self.base_metric_group
+            )
+        )
+        self._retry_enabled = False
+        if output_mode == AsyncFunctionDescriptor.OutputMode.UNORDERED:
+            self._queue = UnorderedStreamElementQueue(capacity, self._raise_exception_if_exists)
+        else:
+            raise NotImplementedError()
+        self._emitter = None
+        self._async_function_runner = None
+        self._exception = None
+
+    def set_output_processor(self, output_processor):
+        self._output_processor = output_processor
+
+    def open(self):
+        self.open_func()
+        self._emitter = Emitter(self._mark_exception, self._output_processor, self._queue)
+        self._emitter.daemon = True
+        self._emitter.start()
+
+        self._async_function_runner = AsyncFunctionRunner(self._mark_exception)
+        self._async_function_runner.daemon = True
+        self._async_function_runner.start()
+
+    def close(self):
+        self.close_func()
+        if self._emitter is not None:
+            self._emitter.stop()
+            self._emitter = None
+
+        if self._async_function_runner is not None:
+            self._async_function_runner.stop()
+            self._async_function_runner = None
```

Comment:
do we need `self._exception = None` here?
In other words, when failover happens, will we reuse `AsyncOperation` object or just create a new one?